tests/test_states.py

```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        """Test that the correct result is returned for E(xp)"""
+        self.logTestName()
+
+        # set quadratic coefficient
+        A = np.zeros([6, 6])
+        A[3, 0] = A[0, 3] = 1
+
+        d = None
+        k = 0
+
+        # prepare a squeezed displaced state
+        self.circuit.reset(pure=self.kwargs['pure'])
+        self.circuit.prepare_displaced_squeezed_state(a, r, phi, 0)
+
+        state = self.circuit.state()
+        mean, var = state.poly_quad_expectation(A, d, k, phi=self.qphi)
+
+        mean_ex, var_ex = self.sample_one_mode_normal_expectations(
+            self.mu, self.cov, lambda X, P, XP: A[3, 0]*XP)
+        self.assertAlmostEqual(mean, mean_ex, delta=self.tol)
+        self.assertAlmostEqual(var, var_ex, delta=self.tol)
+
+    def test_arbitrary_quadratic(self):
+        """Test that the correct result is returned for E(c0 x^2 + c1 p^2 + c2 xp + c3 x + c4 p + k)"""
+        self.logTestName()
+
+        c0 = 1/np.sqrt(2)
+        c1 = 3
+        c2 = 0.53
+        c3 = 1/3.
+        c4 = -1
+
+        # define the arbitrary quadratic
+        A = np.zeros([6, 6])
+        A[0, 0] = 2*c0
+        A[3, 3] = 2*c1
+
+        A[3, 0] = c2
+        A[0, 3] = c2
+
+        # define the linear combination and constant term
+        d = np.array([c3, 0, 0, c4, 0, 0])
+        k = 5
+
+        # prepare a squeezed displaced state
+        self.circuit.reset(pure=self.kwargs['pure'])
+        self.circuit.prepare_displaced_squeezed_state(a, r, phi, 0)
+
+        state = self.circuit.state()
+        mean, var = state.poly_quad_expectation(A, d, k, phi=self.qphi)
+
+        mean_ex, var_ex = self.sample_one_mode_normal_expectations(
+            self.mu, self.cov, lambda X, P, XP: c0*X**2 + c1*P**2 + c2*XP + c3*X + c4*P + k)
+
+        self.assertAlmostEqual(mean, mean_ex, delta=self.tol)
+        self.assertAlmostEqual(var, var_ex, delta=self.tol)
+
+
+class PolyQuadExpectationMultiMode(BaseTest):
+    num_subsystems = 3
+    cutoff = 12
+    hbar = 2
+    qphi = 0.78
+
+    # testing parameters
+    linear_modes = 2
+
+    # define circuit parameters
+    a_list = [0.344+0.123j, 0.432+0.123j, 0.1231-0.543j][:linear_modes]
+    r_list = [0.1065, 0.32, 0.4123][:linear_modes]
+    phi_list = [0.897, 0.31, 0.9643][:linear_modes]
+
+
+    def setUp(self):
+        """Set up."""
+        super().setUp()
+        self.eng, q = sf.Engine(self.num_subsystems, hbar=self.hbar)
+        self.eng.backend = self.backend
+        self.backend.reset(cutoff_dim=self.cutoff)
+
+        if isinstance(self.backend, backends.TFBackend):
+            raise unittest.SkipTest('The poly_quad_expectation method is not yet supported on the TF backend.')
+
+        mu = np.zeros([6])
+        cov = np.zeros([6, 6])
+
+        # construct the vector of means and covariance matrix for each mode
+        # note that we use (x0,p0,x1,p1,...) ordering here.
+        for i, (x, y, z) in enumerate(zip(self.a_list, self.r_list, self.phi_list)):
```

Comment:
Might help for code clarify to use `i, (a, r, phi)`

Reply:
I originally had that, unfortunately the linter complains that we are modifying global variables

Reply:
`(a_, r_, phi_)`?